keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
+updates the new QoSResources field in [NodeStatus](#nodestatus) accordingly,
+making QoS-class resources on the node visible to users and the kube-scheduler.
+
+An admission handler is added into kubelet to validate the QoS-class resource
+request against the resource availability on the node. Pod is rejected if
+sufficient resources do not exist.
+
+Invalid class assignments will cause an error in the container runtime which
+causes the corresponding CRI RuntimeService request (e.g.  RunPodSandbox or
+CreateContainer) to fail with an error. This can happen, despite the admission
+check, e.g. when the requested QoS-class resource is not available on the
+runtime (e.g. kubelet sends the pod or container start request before getting
+an update of changed QoS-class resource availability from the runtime).
+
+A feature gate QoSResources enables kubelet to interpretthe specific pod
+annotations. If the feature gate is disabled the annotations are simply ignored
+by kubelet.
+
+### API server
+
+Input validation of QoS-class resource names and class names, very similar to
+labels is implemented: keys (`QoSResourceName`) and values must be non-empty,
+less than 64 characters long, must start and end with an alphanumeric character
+and may contain only alphanumeric characters, dashes, underscores or dots (`-`,
+`_` or `.`). Also similar to labels, a namespace prefix (FQDN subdomain separated
+with a slash) in the key is allowed (e.g. `vendor/qos-resource`).
+
+### Scheduler
+
+The kube-scheduler will be extended to do node pre-filtering based on the
+QoS-class resource requests of a Pod. If any of the requested QoS-class
+resources are unavailable on a node, that node is not fit. If no nodes can
+satisfy all of the requests the Pod is marked as unschedulable and will be
+scheduled in the future when/if requests can be satisfied by some node.
+
+In principle, scheduling will behave similarly to native and extended
+resources. The kube-scheduler will also take in account the per-class capacity
+on the nodes if that is provided.
+
+### Kubectl
+
+The kubectl describe command will be extended to:
+
+- available QoS-class resources of nodes
+- display QoS-class resource requests of pods
+
+### Container runtimes
+
+Currently, there is support (container-level QoS-class resources) for Intel RDT
+and blockio in CRI-O and containerd runtimes:
+
+- cri-o:
+  - [~~Add support for Intel RDT~~](https://github.com/cri-o/cri-o/pull/4830)
+  - [~~Support for cgroups blockio~~](https://github.com/cri-o/cri-o/pull/4873)
+- containerd:
+  - [~~Support Intel RDT~~](https://github.com/containerd/containerd/pull/5439)
+  - [~~Support for cgroups blockio~~](https://github.com/containerd/containerd/pull/5490)
+
+The design paradigm here is that the container runtime configures the QoS-class
+resources according to a given configuration file. Enforcement on containers is
+done via OCI. User interface is provided through pod and container annotations.
```

Comment:
Does this mean we couldn't specify QoS assignments for networking (not handled by the CRI) or other aspects of a Pod?

Reply:
No. CRI API is only used to carry the information what class a pod or container was assigned to. How it is enforced in the container runtime can be anything. E.g. in the case of network it can hook into the network setup phase of pod sandbox creation (at that time it already knows what "network" class the pod was assigned to)